### PR TITLE
Make tool for go.mod updates

### PR DESCRIPTION
**Pull request by ValarDragon**
## What is the purpose of the change

WIP: python script for recursively updating all of our go.mod's hash committing along the dependency graph.

Most of the code is made via AI tooling.

## Brief Changelog

*(for example:)*
 
  - *The metadata is stored in the blob store on job creation time as a persistent artifact*
  - *Deployments RPC transmits only the blob storage reference*
  - *Daemons retrieve the RPC data from the blob cache*

## Testing and Verifying

*(Please pick one of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added unit test that validates ...*
  - *Added integration tests for end-to-end deployment with ...*
  - *Extended integration test for ...*
  - *Manually verified the change by ...*

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

File: scripts/update_versions.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+  # Iterate over the list of go.mod filepaths
+  for filepath in mod_filepaths:
+    curname = parse_go_mod_name(filepath)
+    # Open the go.mod file in read mode
+    with open(filepath + "/go.mod", 'r') as mod_file:
+      # Read the file contents into a string
+      contents = mod_file.read()
+      
+      # Use a regular expression to match the 'require' block and extract the dependencies
+      dependencies = re.findall(r'require\s+\(\s*(.*?)\s*\)', contents, re.DOTALL)
```
Comment (nicolaslara) [inline scripts/update_versions.py:123]: This assumes no replace directives, which I think is safe to assume.

File: scripts/update_versions.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        print("detected diff relative to main in " + dependency)
+        if check_for_diff_since_last_commit():
+          print("detected diff relative to last commit, committing" + dependency)
+          subprocess.run("git add .".split(" "), capture_output=True)
+          subprocess.run("git commit -m".split(" ") + ["(auto) locking dependency versions for " + dependency])
+          num_commits += 1
+        v = get_go_sum_version()
+        update_dep_versions[dependency] = str(v)
+
+    gitResetStr = "git reset HEAD" + "~"*num_commits
```
Comment (nicolaslara) [inline scripts/update_versions.py:36]: lol. I guess this works, but funny command

File: scripts/update_versions.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    input("going to run git push, hit enter to confirm. Ctrl-C to cancel." + 
+      " (And if so, you likely want to do " + gitResetStr + " to undo the commits)")
```
Comment (nicolaslara) [inline scripts/update_versions.py:38]:
```suggestion
      " (If you cancel, you likely want to do " + gitResetStr + " to undo the commits)")
```

File: scripts/update_versions.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    
+    # Perform a topological sort on the dependency graph and reverse the resulting list
+    sorted_dependencies = list(nx.topological_sort(G))[::-1]
+
+    # now we go commit every go mod file in order, replacing dependencies of prior ones.
+    update_dep_versions = {}
+    num_commits = 0
+    for dependency in sorted_dependencies:
+      # see if we need to change anything per update_dep_versions
+      mod_path = names_to_path[dependency]
```
Comment (nicolaslara) [inline scripts/update_versions.py:24]: can we use the same indent line-length everywhere? (4 spaces). Maybe we want to run this through `black`

File: scripts/update_versions.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+    paths = parse_go_work_paths()
+      # Create a mapping from the paths to the names of the modules
+    path_to_name = {path: parse_go_mod_name(path) for path in paths}
+    names = sorted(path_to_name.values())
+    names_to_path = {path_to_name[path]: path for path in path_to_name}
+    # build dependency graph for our commit order
+    G = build_di_graph(names, paths)
+    
+    # Perform a topological sort on the dependency graph and reverse the resulting list
+    sorted_dependencies = list(nx.topological_sort(G))[::-1]
```
Comment (nicolaslara) [inline scripts/update_versions.py:17]: Cool! this has the side effect of automatically checking for circular dependencies

File: scripts/update_versions.py (hunk 2 above)
Comment (ValarDragon) [reply scripts/update_versions.py:36]: :joy:

File: scripts/update_versions.py (hunk 1 above)
Comment (peterbourgon) [reply scripts/update_versions.py:123]:
```shell
$ find . -name go.mod | xargs grep replace
./go.mod:replace (
./go.mod:       // ToDo: replace the commit hash with v0.30.0-osmo-v14 once the version is tagged
./osmomath/go.mod:replace (
./x/ibc-hooks/go.mod:replace (
./osmoutils/go.mod:replace (
```

File: scripts/update_versions.py (hunk 4 above)
Comment (ValarDragon) [reply scripts/update_versions.py:24]: Good idea, let me setup black in vscode

**Pull request closed without merging**